SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp

```diff
@@ -0,0 +1,476 @@
+{
+}
+
+EPushRPCResult CrossServerRPCService::PushCrossServerRPC(Worker_EntityId EntityId, const RPCSender& Sender,
+														 const PendingRPCPayload& Payload, bool bCreatedEntity)
+{
+	CrossServerEndpoints* Endpoints = CrossServerDataStore.Find(Sender.Entity);
+	Schema_Object* EndpointObject = nullptr;
+	EntityComponentId SenderEndpointId(Sender.Entity, SpatialConstants::CROSSSERVER_SENDER_ENDPOINT_COMPONENT_ID);
+
+	if (!Endpoints)
+	{
+		if (bCreatedEntity)
+		{
+			return EPushRPCResult::EntityBeingCreated;
+		}
+
+		EndpointObject = Schema_GetComponentDataFields(RPCStore->GetOrCreateComponentData(SenderEndpointId));
+		Endpoints = &CrossServerDataStore.Add(Sender.Entity);
+	}
+	else
+	{
+		EndpointObject = Schema_GetComponentUpdateFields(RPCStore->GetOrCreateComponentUpdate(SenderEndpointId));
+	}
+
+	CrossServer::WriterState& SenderState = Endpoints->SenderState;
+
+	TOptional<uint32> Slot = SenderState.Alloc.PeekFreeSlot();
+	if (!Slot)
+	{
+		return EPushRPCResult::DropOverflowed;
+	}
+
+	uint64 NewRPCId = SenderState.LastSentRPCId++;
+	uint32 SlotIdx = Slot.GetValue();
+
+	RPCRingBufferDescriptor Descriptor = RPCRingBufferUtils::GetRingBufferDescriptor(ERPCType::CrossServer);
+	uint32 Field = Descriptor.GetRingBufferElementFieldId(ERPCType::CrossServer, SlotIdx + 1);
+
+	Schema_Object* RPCObject = Schema_AddObject(EndpointObject, Field);
+
+	RPCTarget Target(CrossServerRPCInfo(EntityId, NewRPCId));
+	CrossServer::WritePayloadAndCounterpart(EndpointObject, Payload.Payload, Target, SlotIdx);
+
+	Schema_ClearField(EndpointObject, Descriptor.LastSentRPCFieldId);
+	Schema_AddUint64(EndpointObject, Descriptor.LastSentRPCFieldId, SenderState.LastSentRPCId);
+
+	CrossServer::RPCKey RPCKey(Sender.Entity, NewRPCId);
+	CrossServer::SentRPCEntry Entry;
+	Entry.Target = Target;
+	Entry.SourceSlot = SlotIdx;
+
+	SenderState.Mailbox.Add(RPCKey, Entry);
+	SenderState.Alloc.CommitSlot(SlotIdx);
+
+	return EPushRPCResult::Success;
+}
+
+void CrossServerRPCService::AdvanceView()
+{
+	const FSubViewDelta& SubViewDelta = SubView->GetViewDelta();
+	for (const EntityDelta& Delta : SubViewDelta.EntityDeltas)
+	{
+		switch (Delta.Type)
+		{
+		case EntityDelta::UPDATE:
+		{
+			for (const ComponentChange& Change : Delta.ComponentUpdates)
+			{
+				ComponentUpdate(Delta.EntityId, Change.ComponentId, Change.Update);
+			}
+			break;
+		}
+		case EntityDelta::ADD:
+			PopulateDataStore(Delta.EntityId);
+			break;
+		case EntityDelta::REMOVE:
+		case EntityDelta::TEMPORARILY_REMOVED:
+			CrossServerDataStore.Remove(Delta.EntityId);
+			RPCStore->PendingComponentUpdatesToSend.Remove(
+				EntityComponentId(Delta.EntityId, SpatialConstants::CROSSSERVER_SENDER_ENDPOINT_COMPONENT_ID));
+			RPCStore->PendingComponentUpdatesToSend.Remove(
+				EntityComponentId(Delta.EntityId, SpatialConstants::CROSSSERVER_RECEIVER_ACK_ENDPOINT_COMPONENT_ID));
+			if (Delta.Type == EntityDelta::TEMPORARILY_REMOVED)
+			{
+				PopulateDataStore(Delta.EntityId);
```

Comment:
Does TEMPORARILY_REMOVED mean the data won't be in the subview?

Reply:
It means that the entity flashed out/in interest, so it disappeared for a while, then reappeared.